tensorflow/python/ops/init_ops.py

```diff
@@ -340,3 +341,45 @@ def __call__(self, shape, dtype=dtypes.float32, partition_info=None):
     """Generate a tensor used to initialize a variable."""
     return random_ops._random_walk(shape, self._nonlinearity, dtype,
                                    seed=self._seed)
+
+
+def orthogonal_initializer(gain=1.0, dtype=dtypes.float32, seed=None):
+  """Returns an initializer that generates an orthogonal rotation matrix.
+
+  Initialize a tensor using an orthogonal matrix.
+
+  Args:
+    gain: multiplicative factor to apply to the orthogonal matrix
+    dtype: The type of the output.
+    seed: A Python integer. Used to create random seeds. See
+      [`set_random_seed`](../../api_docs/python/constant_op.md#set_random_seed)
+      for behavior.
+
+  Returns:
+    An initializer that generates orthogonal tensors
+
+  Raises:
+    ValueError: if `dtype` is not a floating point type.
+  """
+  def _initializer(shape, dtype=_assert_float_dtype(dtype), partition_info=None):
+    # Flatten the input shape with the last dimension remaining its original shape so it works for conv2d
+    num_rows = 1
+    for dim in shape[:-1]:
+      num_rows *= dim
+    num_cols = shape[-1]
+    flat_shape = (num_rows, num_cols)
+
+    if gain == 'relu':
+      _gain = math.sqrt(2)
+    else:
+      _gain = gain
+
+    # Generate a random matrix
+    a = random_ops.random_normal(flat_shape, dtype=dtype, seed=seed)
+    # Compute the svd
+    _, u, v = linalg_ops.svd(a, full_matrices=False)
+    # Pick the right singular value decomposition
+    q = u if num_cols < num_rows else v
+    return _gain * array_ops.reshape(q, shape)
```

Comment:
I think u is (num_rows x num_rows), and v is (num_cols x num_cols). tf.reshape requires the old shape and the new shape to be identical. So there is no guarantee this would work.